File: spfcheck.pl
#!/usr/bin/perl

use strict;
use warnings;
use Mail::SPF::Query;

my $query = new Mail::SPF::Query (ip => $ENV{'TCPREMOTEIP'}, sender=>$ENV{'SMTPMAILFROM'}, helo=>$ENV{'SMTPHELOHOST'}, trusted=>1, guess=>1);

my ($result,           # pass | fail | softfail | neutral | none | error | unknown [mechanism]
    $smtp_comment,     # "please see http://spf.pobox.com/why.html?..."  when rejecting, return this string to the SMTP client
    $header_comment,   # prepend_header("Received-SPF" => "$result ($header_comment)")
    $spf_record,       # "v=spf1 ..." original SPF record for the domain
   ) = $query->result2();

# Relay- and Trustclients pass right through
if ( (defined($ENV{'RELAYCLIENT'})) ||
     (defined($ENV{'TRUSTCLIENT'})) ) {
  print "\n";
  exit 0;
}

# Reject the client if SPF check returns "fail"
if ($result eq 'fail') {
  print STDERR getppid()." SPF check failed for $ENV{'SMTPMAILFROM'} !\n";
  print STDERR getppid()." Mail recipient would have been $ENV{'SMTPRCPTTO'}\n";
  print "E451 SPF check failed: $smtp_comment\n";
} else {
  print "\n";
}

exit;

File: spfcheck.go
package main

import "fmt"
import "net"
import "blitiri.com.ar/go/spf"
import "os"
import "strings"
import "time"

func main() {
	if env_defined("RELAYCLIENT") ||
		env_defined("TRUSTCLIENT") {
		fmt.Println()
		os.Exit(0)
	}

	if !env_defined("TCPREMOTEIP") {
		fmt.Println()
		os.Exit(0)
	}

	var smtpmailfrom string = os.Getenv("SMTPMAILFROM")
	var addrparts []string = strings.Split(smtpmailfrom, "@")

	ip := net.ParseIP(os.Getenv("TCPREMOTEIP"))
	r, _ := spf.CheckHost(ip, addrparts[1])

	if (r == "fail") {
		fmt.Fprintf(os.Stderr, "%d SPF check failed for %s!\n", os.Getppid(), smtpmailfrom)
		fmt.Fprintf(os.Stderr, "%d Mail recipient would have been %s\n", os.Getppid(), os.Getenv("SMTPRCPTTO"))
		time.Sleep(5)
		fmt.Println("E451 SPF check failed")
	} else {
		fmt.Println()
	}

	os.Exit(0)
}

func env_defined(key string) bool {
	value, exists := os.LookupEnv(key)
	_ = value

	return exists
}
